[MRG+1] Vectorize connectivity

salma1601 opened the pull request:
try to fix #1233, #1229, #1237

AlexandreAbraham commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -231,6 +231,42 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+    p = (sqrt(8 * n + 1) - 1.) / 2
+    if p > floor(p):
+        raise ValueError(
+            "Vector of unstuitable shape {0} can not be transformed to "
```
typo

AlexandreAbraham commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -231,6 +231,42 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+    p = int(p)
+    mask = np.tril(np.ones((p, p))).astype(np.bool)
+    sym = np.zeros(vec.shape[:-1] + (p, p))
+    sym[..., mask] = vec.copy()  # avoid side effects
```
Why would there be a side effect?

salma1601 replied:
I thought changing the diagonal of sym may change vec but I am wrong

AlexandreAbraham commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -231,6 +231,42 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+    mask = np.tril(np.ones((p, p))).astype(np.bool)
+    sym = np.zeros(vec.shape[:-1] + (p, p))
+    sym[..., mask] = vec.copy()  # avoid side effects
+    sym.swapaxes(-1, -2)[..., mask] = vec
```
Can you put a comment saying that this line is to fill the upper triangle?

salma1601 replied:
OK. Does it make sense to make vec_to_sym private? I don't think it will be used apart from here

GaelVaroquaux replied:
> OK. Does it make sense to make vec_to_sym private? I don't think it will be used apart from here

I think that people (methods developers) might use it to build their own
algorithms. As nilearn is also for methods developers, I think that it
makes sense to expose it publicaly.

AlexandreAbraham commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -231,6 +231,42 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+
+    # Rescale diagonal terms
+    mask.flat[::p + 1] = False
+    mask = np.logical_not(mask + mask.T)  # True for diagonal terms
```
Overly complicated.
Simpler:
```python
mask.fill(False)  # Avoid memory re-allocation
np.fill_diagonal(mask, True)
```

salma1601 commented:
can I use this PR to fix also #1237 ? I don't know if this is a good practice

GaelVaroquaux commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -231,6 +231,52 @@ def sym_to_vec(symmetric, discard_diagonal=False):
     return symmetric[..., tril_mask] / scaling[tril_mask]
 
 
+def vec_to_sym(vec):
```
The name is a bit hard to understand out of context. How about "vec_to_sym_matrix". A bit longer, but much more explicit.

GaelVaroquaux commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -284,13 +330,17 @@ class ConnectivityMeasure(BaseEstimator, TransformerMixin):
             "covariance", "precision"}, optional
         The matrix kind.
 
+    vectorize : bool, optional
```
I'd rather have this done inside the ConnectivityMeasure object (in the transform method), rather than the function. It would minimize surprise with regards to what this function does.

GaelVaroquaux replied:
Actually, this is what's done. It's just that the docstring is misleading.

salma1601 replied:
you mean I should put the `vectorize` doc not in the class ?

GaelVaroquaux replied:
I mean that I didn't read the diff properly :$. Sorry.

The description of the argument could be a tiny bit longer, that said, explaining what you mean by vectorizing.

GaelVaroquaux commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -390,4 +443,49 @@ def transform(self, X):
+            self.mean_ = connectivities.mean(axis=0)
+
+        if self.vectorize:
+            connectivities = sym_to_vec(connectivities)
```
I would also rename sym_to_vec to sym_matrix_to_vec

GaelVaroquaux commented:
The example should be updated.

salma1601 commented:
>The example should be updated.

Th

KamalakerDadi commented on doc/whats_new.rst:
```diff
@@ -254,7 +254,7 @@ New features
      :class:`nilearn.connectome.ConnectivityMeasure` can be useful for
      computing functional connectivity matrices.
-   - The function :func:`nilearn.connectome.sym_to_vec` in same module
+   - The function :func:`nilearn.connectome.sym_matrix_to_vec` in same module
```
you may need to revert it the same ```sym_to_vec``` and instead add in section 0.2.6 about these API changes ```sym_to_vec``` to ```sym_matrix_to_vec``` and new features ```vectorize``` and ```inverse_transform``` and example renaming with this PR. Whatever might be interesting to know about this PR can be added in 0.2.6

KamalakerDadi replied:
Now that 0.2.6 is released. Can you please update ?
ping @GaelVaroquaux @AlexandreAbraham next going to be 0.2.6.1 or 0.2.7 ?

salma1601 replied:
You mean I add a new section 0.2.6.1 and put the changes in it

KamalakerDadi replied:
Yes, because changes you made in a section is released. New section should be added and included in what ever might be useful to users to refer.

AlexandreAbraham replied:
> ping @GaelVaroquaux @AlexandreAbraham next going to be 0.2.6.1 or 0.2.7 ?

You can put either of these or even 2.X.X. The "size" of the revision is decided at release time depending if backward compatibility is broken or not.

KamalakerDadi commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,259 @@
+Functional connectivity kinds for group analysis of connectomes
+===============================================================
+
+This example compares different kinds of functional connectivity between
```
You can keep it if you are comfortable with saying 'kind' but 'coefficients' might be more relevant I guess ?

For example, 'compares different measures/estimations of functional connectivity coefficients between .....' ?

GaelVaroquaux replied:
I would say "This example compares different types of functional connectivity matrices ...

KamalakerDadi commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,259 @@
+print('Data has {0} ADHD subjects.'.format(len(adhd_subjects)))
+
+###############################################################################
+# ROI-to-ROI correlations of ADHD people
```
people doesn't sounds good to me. Can we write people -> subjects ?

AlexandreAbraham replied:
Individuels or participant is better than subject.

KamalakerDadi commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,259 @@
+
+###############################################################################
+# The simpler and most commonly used kind of connectivity is correlation. It
+# models the total connectivity between pairwise ROIs. We can estimate it using
```
What it means by total connectivity ?

salma1601 replied:
I meant total connectivity as opposed to partial connectivity

KamalakerDadi commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,259 @@
+
+###############################################################################
+# Firstly we Z-Fisher transform the correlation matrices. Beforehand, we
+# reshape each matrix to a vector while discaring their diagonals.
```
typo discaring -> discarding

KamalakerDadi commented:
In the example: May be I missed something ? Why aren't we displaying statistically tested mean connectivity matrices for partial and tangent. Why only for correlation one ?

KamalakerDadi commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,259 @@
+
+###############################################################################
+# `tangent_matrices` are complex features that do not reflect brain connections
+# directly. An additionnal step is needed to recover connectivity.
```
typo additionnal -> additional

KamalakerDadi commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,259 @@
+###############################################################################
+# `tangent_matrices` are complex features that do not reflect brain connections
+# directly. An additionnal step is needed to recover connectivity.
+connectivity_matrices = tangent_measure.inverse_transform(tangent_matrices)
```
Correct me if I am wrong ? inverse_transform is basically useful to invert vectors to matrices. Isn't it ? But, here we are inverting again to matrices -> matrices but not vectorized one.

salma1601 replied:
Yes. In all cases `inverse_transform` gives individual connectivity matrices, whether or not vectorization happened. So if it was not asked for vectorization, `inverse_transform` does nothing for `correlation` and `partial correlation`. However for `tangent`, what I am calling `tangent_matrices` are not true connectivity matrices because for instance they can have negative values on the diagonal. The only way to recover the `connectivity_matrices` is to inverse transform.
I can't figure out a way to be more pedagogic on this mess, so if you have suggestions...

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -231,6 +231,73 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+def vec_to_sym_matrix(vec, diagonal=None):
+    """Return the symmetric array given its flattened lower triangular part.
+    Acts on the last dimension of the array if not 1-dimensional.
+    Diagonal can be encompassed in vec or given seperatly. In both cases, note
```
seperatly -> separately

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -231,6 +231,73 @@ def sym_to_vec(symmetric, discard_diagonal=False):
 
 
+def vec_to_sym_matrix(vec, diagonal=None):
+    """Return the symmetric array given its flattened lower triangular part.
```
May be array to matrix ?

salma1601 replied:
I prefer array because it works also for dimension more than 2

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -231,6 +231,73 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+    Returns
+    -------
+    sym : numpy.ndarray, shape (..., p, p) or (..., p + 1, p + 1) if diagonal
+        is given seperately.
```
seperately -> separately

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -231,6 +231,73 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+    sym : numpy.ndarray, shape (..., p, p) or (..., p + 1, p + 1) if diagonal
+        is given seperately.
+        The output symmetric array.
+
```
Can you add See Also section to ```sym_matrix_to_vec``` ?

KamalakerDadi commented:
Hi @salma1601, Any progress on this ? Thanks

salma1601 commented:
Hi @KamalakerDadi ,
Sorry, but I didn't finish yet because I didn't have time to work on the statistical tests for the `tangent` kind.
I am not sure that student t-test is well suited to the vectors output of `tangent`, I think it will be a chisquare test. I want to work on that to be sure, and I don't have time right now.
Do you think it would be better if I remove the test part, address the other comments, and do another PR to add the tests when I am ready ?

GaelVaroquaux commented:
> I am not sure that student t-test is well suited to the vectors output of
> tangent,

Why? They should be normaly-distributed under the null (see eg my paper).

KamalakerDadi commented:
>Do you think it would be better if I remove the test part, address the other comments, and do another PR to add the tests when I am ready ?

Quickly looking at your files, you already have tests included. Doesn't that looks good ?

Let me also go through once the complete PR thoroughly.

salma1601 commented:
>Quickly looking at your files, you already have tests included

I didn't mean testing the API, I meant statistical tests in the example, when I am using `scipy.stats.ttest_1samp` to extract significant connections.
We can do these statistical tests on the Z-Fisher transformed `correlation` or `partial correlation` coefficients, because there is some theory behind. But for the `tangent` kind the theory is not the same, and the statistical tests would *rather* be `scipy.stats.chisquare`. I say *rather* because you see I am not sure, and I need to look into this but I have no time for the moment.

salma1601 commented:
>Why? They should be normaly-distributed under the null (see eg my paper)

In your paper I think the subject displacements are *assumed* to follow a normal distribution. From what I understand, to  do the testing we need to have a Central Limit Theorem that proves the asymptotic normality. Actually such theorem exist for the `tangent` mean, but those theoretical papers point to chi square tests.

Sorry for not being more accurate, I really need to work on that and write things down to be sure ...

salma1601 commented:
I am going to remove the statistical tests and address the remaining issues.
The statistical tests require additional work for `tangent` coef that is out of scope of this PR

GaelVaroquaux commented:
> I meant total connectivity as opposed to partial connectivity

The technical word is "marginal". I would say "full (marginal)
connectivity", as to make it understandable for non-technical people.

Note that marginal has been used in neuroimaging publications.

GaelVaroquaux commented:
I disagree with you that tangent matrices cannot and shouldn't be
interpreted. I think that they very much should. They are the natural
parametrization of variability on the covariance matrices.

GaelVaroquaux commented:
> I prefer array because it works also for dimension more than 2

I still prefer matrices: it's the vocabulary that people will have in
mind.

salma1601 commented:
>I still prefer matrices: it's the vocabulary that people will have in mind

OK

>I disagree with you that tangent matrices cannot and shouldn't be interpreted. I think that they very much should. They are the natural parametrization of variability on the covariance matrices.

Do you mean I should plot them and interpret them in the example, rather than the individual covariances ?

GaelVaroquaux commented:
> Do you mean I should plot them and interpret them in the example, rather than
> the individual covariances ?

I think so. Just like in my old stroke paper.

salma1601 commented:
>I think so. Just like in my old stroke paper.

Ok, I though it will be too much technical, but Ok, happy to do it !

KamalakerDadi commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,218 @@
+correlation_measure = ConnectivityMeasure(kind='correlation')
+
+###############################################################################
+# From the list of ROIs time-series for all subjects, the `correlation_measure`
```
all subjects -> adhd subjects. Since you are fitting on adhd subjects not on pooled subjects ?

KamalakerDadi commented:
@salma1601 I am unable to run the example ? receiving a ```TypeError```. Are you able to run example ?

KamalakerDadi commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,218 @@
+    plot_matrix(matrix, 'correlation, subject {0}'.format(n))
+
+plotting.plot_connectome(mean_correlation_matrix, msdl_coords,
+                         title='mean correlation')
```
can you be explicit in title="mean correlation over 13 ADHD subjects" ?

KamalakerDadi commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,218 @@
+    plot_matrix(matrix, 'partial, subject {0}'.format(n))
+
+plotting.plot_connectome(partial_correlation_measure.mean_, msdl_coords,
+                         title='mean partial correlation')
```
same comment as above more explicit title "mean partial correlation over 13 ADHD subjects" ?

KamalakerDadi commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,218 @@
+    plot_matrix(matrix, 'tangent variability, subject {0}'.format(n))
+
+plotting.plot_connectome(tangent_measure.mean_, msdl_coords,
+                         title='mean tangent')
```
Same here more explicit title "mean tangent matrices over 13 ADHD subjects"?

KamalakerDadi commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,218 @@
+
+mean_scores = []
+for kind in kinds:
+    cv_scores = cross_val_score(LinearSVC(), connectivity_biomarkers[kind],
```
May be we need LinearSVC(random_state=0) ?

KamalakerDadi commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,218 @@
+
+classes = ['{0}{1}'.format(site_name, adhd_label)
+           for site_name, adhd_label in zip(site_names, adhd_labels)]
+cv = StratifiedKFold(classes, n_folds=3)
```
Shouldn't it be given adhd_labels instead of classes ? Correct me if I am wrong ?

salma1601 replied:
I am giving both `adhd_labels` and `site_names` to constraint also by sites the splitting of data, no ? For instance I don't want ADHD patients to be exclusively from NYU site in one fold, and no NYU patients in others, because the matrices vary a lot between sites (actually they vary more across sites than between ADHD and controls ! that's why I think it is important to conserve the same sites proportion in each fold)
Do you mean I should rename `classes` to something like `subjects_labels` ?

GaelVaroquaux replied:
This is great.

I think that you should expand a bit your comment so that people understand better what you are doing.

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -231,6 +231,77 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+
+    Parameters
+    ----------
+    vec : numpy.ndarray, shape (..., p * (p + 1) /2)
```
What is p ?

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -231,6 +231,77 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+    -------
+    sym : numpy.ndarray, shape (..., p, p) or (..., p + 1, p + 1) if diagonal
+        is given seperately.
+        The output symmetric array.
```
symmetric "matrix" ?

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -231,6 +231,77 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+    nilearn.connectome.sym_matrix_to_vec
+    """
+    n = vec.shape[-1]
+    # solve p * (p + 1) / 2 = n subj. to p > 0
```
are these commented (#) calculations necessary ?

salma1601 replied:
I think they help explaining what's done, because otherwise the value of `p` comes from nowhere.
You disagree ?

KamalakerDadi replied:
I agree, may be we could add this in documentation. What do you think ? 
Or
 We could leave as it is.

salma1601 replied:
I think it would be better add a function `_compute_matrix_shape` and document it correctly

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -231,6 +231,77 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+        n_columns = p
+        diagonal_offset = 0
+    else:
+        if diagonal.shape[:-1] != first_shape or diagonal.shape[-1] != p + 1:
```
diagonal should be given and should be same as length of features or samples ?

I didn't understand the condition here, ```diagonal.shape[-1] != p + 1```

salma1601 replied:
>diagonal should be given

`vec_to_sym_matrix` is meant to be the reverse to `sym_matrix_to_vec`. So if you have discarded the diagonal in `sym_matrix_to_vec`, you need to provide it separately to reconstruct the symmetric matrix. For instance this can be useful for correlation matrices for which we know the diagonal is 1.

> should be same as length of features or samples 
if you provide it, it should have the right shape, that is for n-dimensional array: same as `vec` until n-1, and matching the matrix shape on the last dimension

salma1601 replied:
>I didn't understand the condition here, `diagonal.shape[-1] != p + 1`

I have to correct the doc, but `p + 1` is correct here.
Your questions tell me my code is unclear and I have to rework it. May be I can add a little function that computes the shape of the symmetric matrix from the given vector, and also add a section example to illustrate how it works: if you enter a vector that does not include the diagonal in it, the computed shape from the equation is less than the true matrix shape
```Python
sym = np.array([[1., .3, .1], [.3, 1., .5], [.1, .5, 1.]])
vec = sym_matrix_to_vec(sym, discard_diagonal=True)
# Recover the matrix shape
n = np.shape(vec)[-1]
p = (sqrt(8 * n + 1) - 1.) / 2
n_columns = int(p) + 1
print n_columns
```
3
```Python
diagonal = np.ones(n_columns,) / sqrt(2)
sym2 = vec_to_sym_matrix(vec, diagonal)
print sym2
```
outs
```Python
[[ 1.   0.3  0.1]
 [ 0.3  1.   0.5]
 [ 0.1  0.5  1. ]]
```

KamalakerDadi replied:
I don't think on the code basis but documentation. You are explaining quite well and I felt some important points are missing in the doc. 

For example, "So if you have discarded the diagonal in sym_matrix_to_vec, you need to provide it separately to reconstruct the symmetric matrix. For instance this can be useful for correlation matrices for which we know the diagonal is 1."

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -357,12 +436,14 @@ def transform(self, X):
         Returns
         -------
-        output : numpy.ndarray, shape (n_samples, n_features, n_features)
+        output : numpy.ndarray, shape (n_subjects, n_features, n_features)
```
What do you change to n_subjects ?

KamalakerDadi replied:
May be we should decide to be consistent with n_samples or n_subjects ?

salma1601 replied:
because there is the number of the timeseries scans, which is n_samples, and the number of subjects

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -284,13 +358,17 @@ class ConnectivityMeasure(BaseEstimator, TransformerMixin):
         The matrix kind.
 
+    vectorize : bool, optional
+        Vectorize or not the connectivities.
```
Can you write description a bit longer ? May be adding few important lines from function to here.

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -284,13 +358,17 @@ class ConnectivityMeasure(BaseEstimator, TransformerMixin):
     `mean_` : numpy.ndarray
-        The mean connectivity for the tangent kind.
+        The mean connectivity across subjects. Note that for 'tangent' kind,
+        averaging procedure is not the trivial one.
```
So which means you are not returning mean_ for kind='tangent' ? May be you could be explicit here ?

salma1601 replied:
Ok

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -390,4 +471,49 @@ def transform(self, X):
+        self._check_fitted()
+
+        connectivities = np.array(connectivities)
+        if self.vectorize:
```
I didn't understand why converting to symmetric matrices is linked to ```self.vectorize``` ? May be I am missing some logic here ?

salma1601 replied:
>I didn't understand why converting to symmetric matrices is linked to `self.vectorize`

because we only vectorize if `self.vectorize=True`, otherwise we already have matrices, no ?

KamalakerDadi replied:
Ok I missed this logic. Thanks.

What would happen if I haven't selected ```vectorize=True``` and then trying to ```inverse_transform``` here ?

salma1601 replied:
>What would happen if I haven't selected vectorize=True and then trying to inverse_transform here ?

for `kind="correlation"` or `kind="partial correlation"`, nothing happens
for `kind="tangent"`, the covariance matrices are reconstructed. The output of `transform` for tangent are not connectivity matrices, but deviations from a group connectivity matrix. They have a different *nature* than the usual connectivity matrices, for instance they can have negative values on their diagonal, while this never happen in covariance, correlation, partial correlation matrices

salma1601 replied:
I feel like I also have to explicit more the doc here ;)

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -390,4 +471,49 @@ def transform(self, X):
+             The corresponding connectivity matrices.
+             If kind is 'tangent', the covariance matrices are reconstructed.
+        """
+        self._check_fitted()
```
May be this ```_check_fitted``` is also essential at ```transform``` level too ? What do you think ?

salma1601 replied:
Ok I add it to `transform`

salma1601 commented:
>@salma1601 I am unable to run the example ? receiving a TypeError. Are you able to run example ?

I run it without problem

KamalakerDadi commented on nilearn/connectome/tests/test_connectivity_matrices.py:
```diff
@@ -322,15 +324,74 @@ def test_geometric_mean_errors():
+    assert_array_almost_equal(sym_matrix_to_vec(sym, discard_diagonal=True),
+                              vec)
+
+    # Check vec_to_sym_matrix is the inverse function of sym_matrix_to_vec
```
May be need to shift this tests to ```vec_to_sym_matrix``` because you are testing in ```sym_matrix_to_vec``` ?

salma1601 commented:
@KamalakerDadi I tried addressing all your comments, thank you very much for the thorough review.
I changed a little bit the `vec_to_sym_matrix` code to make it more intuitive.
I add a version added 0.2.6.1 for this function or not ?

KamalakerDadi commented:
>I add a version added 0.2.6.1 for this function or not ?

Yes, please.

KamalakerDadi commented:
CircleCI failure seems a genuine one. Some where you missed the indendation which made sphinx unhappy. May be I guess in whats_new ?

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -390,4 +492,73 @@ def transform(self, X):
+                if self.kind in ['correlation', 'partial correlation']:
+                    diagonal = np.ones(diagonals_shape) / sqrt(2.)
+                elif self.kind in ['covariance', 'precision']:
+                    warnings.warn("diagonal values have been discared and are "
```
Can you check typos in this warning message ? discared -> discarded unkown-> unknown

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -390,4 +492,73 @@ def transform(self, X):
+                    diagonal[:] = np.nan
+                else:
+                    raise ValueError("diagonal values has been discarded and "
+                                     "are unkown for tangent kind, can not "
```
unkown-> unknown

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -390,4 +492,73 @@ def transform(self, X):
+                elif self.kind in ['covariance', 'precision']:
+                    warnings.warn("diagonal values have been discared and are "
+                                  "unkown for {0} kind, replacing them with "
+                                  "Nan.".format(self.kind))
```
You are filling discarded diagonal with Nan if user hasn't provided diagonal explicitly. Do you think it is Ok ?
Why not we raise an error here ?

salma1601 replied:
Ok, I changed this.
Originally I raised error only for tangent because if diagonal is unknown we can not reconstruct any element of the covariance matrices, while for 'covariance' or 'precision' kind we loose only the diagonal. But it's complicating things with no profit, so I am raising error for all

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -390,4 +492,73 @@ def transform(self, X):
+                             'transform().'.format(self.__class__.__name__)
+                             )
+
+    def inverse_transform(self, connectivities):
```
Is there any option here to submit my diagonal explicitly like in ```vec_to_symmetric_matrix``` ?

KamalakerDadi commented on nilearn/connectome/tests/test_connectivity_matrices.py:
```diff
@@ -322,15 +325,74 @@ def test_geometric_mean_errors():
+    assert_array_almost_equal(sym_matrix_to_vec(sym, discard_diagonal=True),
+                              vec)
+
+    # when diagonal is included
```
This test also tests with input as list. Isnt't it ?

salma1601 replied:
yes, I am adding a comment

KamalakerDadi commented on nilearn/connectome/tests/test_connectivity_matrices.py:
```diff
@@ -322,15 +325,74 @@ def test_geometric_mean_errors():
+    syms = vec_to_sym_matrix(vecs)
+    assert_array_almost_equal(sym_matrix_to_vec(syms), vecs)
+
+    # when diagonal is given seperately
```
same here, tests with lists too, right ?

salma1601 replied:
yes

KamalakerDadi commented on nilearn/connectome/tests/test_connectivity_matrices.py:
```diff
@@ -421,3 +483,93 @@ def test_connectivity_measure_outputs():
+    # Check inverse transformation
+    kinds.remove('tangent')
+    for kind in kinds:
+        # without vectorization
```
can you add few more details here ? I understood that you are returning the same connectivities with vectorize=False.

salma1601 replied:
done

KamalakerDadi commented on nilearn/connectome/tests/test_connectivity_matrices.py:
```diff
@@ -421,3 +483,93 @@ def test_connectivity_measure_outputs():
+
+        # with vectorization
+        # if diagonal has not been discarded
+        conn_measure = ConnectivityMeasure(kind=kind, vectorize=True)
```
do you think these blocks are necessary ? Basically same blocks are repeated below, I guess. Is it not ?

salma1601 replied:
you mean I do a for loop ? the problem is I need the values of connectivities without vectorization for my second check. I am not sure I get your idea

KamalakerDadi commented:
Few more comments again from my side. Otherwise, LGTM.

Can somebody has second review please ?

salma1601 commented on nilearn/connectome/__init__.py:
```diff
@@ -5,7 +5,8 @@
 
 from .connectivity_matrices import (sym_matrix_to_vec, vec_to_sym_matrix,
-                                    ConnectivityMeasure)
+                                    ConnectivityMeasure, cov_to_corr,
```
Yes I rebased but when using mergetool to fix conflicts I skipped accidentally this one

KamalakerDadi commented:
I think CircleCI failure is not related to your PR. We are trying to address that in PR #1312

KamalakerDadi commented:
Any second round of reviews on this ?

AlexandreAbraham commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -216,21 +216,100 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+    sym = np.zeros(first_shape + (n_columns, n_columns))
+
+    # Fill upper triangular part
+    mask = np.tril(np.ones((n_columns, n_columns))).astype(np.bool)
```
```python
np.tril(np.ones((n_columns, n_columns), dtype=bool))
```

AlexandreAbraham replied:
Here you should be able to put k=(diagonal is not None) (or minus this) and fill the diagonal at line 307. This allows you to get rid of the ugly part of this function that copy all the data.

AlexandreAbraham commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -216,21 +216,100 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+
+    sym = np.zeros(first_shape + (n_columns, n_columns))
+
+    # Fill upper triangular part
```
Lower.

AlexandreAbraham commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -216,21 +216,100 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+    mask = np.tril(np.ones((n_columns, n_columns))).astype(np.bool)
+    sym[..., mask] = vec
+
+    # Fill lower triangular part
```
Upper

AlexandreAbraham commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -216,21 +216,100 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+    sym[..., mask] = vec
+
+    # Fill lower triangular part
+    sym.swapaxes(-1, -2)[..., mask] = vec
```
... or you can transpose the mask ;).

salma1601 replied:
This didn't give me exactly what I wanted
```Python
mask = np.tril(np.ones((3, 3), dtype=bool))
sym = np.array([[4, 0, 0], [3, 5, 0], [2, 7, 1]])
print sym[mask.T], sym.T[mask]
```
gives [4 0 5 0 0 1] [4 0 0 5 0 1]

AlexandreAbraham commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -216,21 +216,100 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+    ----------
+    vec : numpy.ndarray, shape (..., n_columns * (n_columns + 1) /2) or
+        (..., (n_columns - 1) * n_columns / 2) if diagonal is given seperately.
+        The input array.
```
My opinion is that most people will want to be able to feed this function lower triangular parts of matrices and one value (0) for the diagonal. Don't you think?

salma1601 replied:
Yes, but do you suggest I put it to zero by default ?

AlexandreAbraham commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -216,21 +216,100 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+        vec_with_diagonal = np.zeros(first_shape + (n + n_columns,))
+        vec_with_diagonal[..., diagonal_mask] = diagonal
+        vec_with_diagonal[..., np.logical_not(diagonal_mask)] = vec
+        return vec_to_sym_matrix(vec_with_diagonal)
```
Reallocating the whole data just to fill one diagonal? This is not necessary and can be done further.

KamalakerDadi commented:
Travis failure seems like a genuine one. Can you please have a look at it ?

salma1601 commented:
@KamalakerDadi Tests are passing !

KamalakerDadi commented:
>@KamalakerDadi Tests are passing !

Great. Thanks.

@GaelVaroquaux @AlexandreAbraham Can you give a final call ?

mrahim commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -193,7 +193,7 @@ def _geometric_mean(matrices, init=None, max_iter=10, tol=1e-7):
     return gmean
 
 
-def sym_to_vec(symmetric, discard_diagonal=False):
```
I'm wondering if it's okay to remove  `sym_to_vec` completely. Could we put  a deprecation instead ?

salma1601 replied:
Done

mrahim commented:
Could you squash your commits a bit please ? :)

salma1601 commented:
squashing done

GaelVaroquaux commented on doc/whats_new.rst:
```diff
@@ -1,3 +1,32 @@
+0.3.1
```
If I merge this today (which is probable) we should have this as 0.3.0 beta2

GaelVaroquaux commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,222 @@
+"""
+Functional connectivity kinds for group analysis of connectomes
```
I would replace "kind" by "matrices".

GaelVaroquaux commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,222 @@
+# Load ADHD dataset and MSDL atlas
+# --------------------------------
+
+###############################################################################
```
We could merge the two comment blocks above

GaelVaroquaux commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,222 @@
+msdl_coords = msdl_data.region_coords
+n_regions = len(msdl_coords)
+print('MSDL has {0} ROIs, part of the following networks:\n{1}.'.format(
+    n_regions, msdl_data.networks))
```
Missing space here

GaelVaroquaux commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,222 @@
+# Region signals extraction
+# -------------------------
+
+###############################################################################
```
Could you merge the 2 comment blocks

GaelVaroquaux commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,222 @@
+from nilearn import input_data
+
+masker = input_data.NiftiMapsMasker(
+    msdl_data.maps, resampling_target="maps", t_r=2.5, detrend=True,
```
Is there a reason that we are not using the default for resampling_target?

salma1601 replied:
Ok, I think we should also fix the doc
https://github.com/nilearn/nilearn/blob/master/nilearn/input_data/nifti_maps_masker.py#L87

GaelVaroquaux commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,222 @@
+# yet it is less dense than correlations graph.
+
+###############################################################################
+# Which connectivity kind is more powerful? A classification answer
```
As a title here, I would suggest "What kind of connectivity is most powerful for classification?"

GaelVaroquaux commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,222 @@
+# Which connectivity kind is more powerful? A classification answer
+# -----------------------------------------------------------------
+
+###############################################################################
```
Could you merge the comment blocks

GaelVaroquaux commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -216,21 +249,97 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+    Diagonal can be encompassed in vec or given separately. In both cases, note
+    that diagonal elements are multiplied by sqrt(2).
+
+    .. versionadded:: 0.2.6.1
```
Now it's 0.3 :)

GaelVaroquaux commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -216,21 +249,97 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+        The input array.
+
+    diagonal : numpy.ndarray, shape (..., n_columns), optional
+        The diagonal array to be stacked to vec.
```
It's unclear to me what is done if diagonal is not given: it is assumed to be included in "vec"? Could you specify this?

salma1601 replied:
Yes if the diagonal is not given it is extracted from `vec`
Ok I will explicit

GaelVaroquaux commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -357,14 +483,19 @@ def transform(self, X):
+        output : numpy.ndarray, shape (n_subjects, n_features, n_features) or \
+                 (n_subjects, n_features * (n_features + 1) / 2) if vectorize \
+                 is set to True.
+            The transformed individual connectivities, as matrices or vectors.
```
Bad indentation in the docstring.

GaelVaroquaux commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -390,4 +521,69 @@ def transform(self, X):
+                             )
+
+    def inverse_transform(self, connectivities, diagonal=None):
+        """Returns connectivity matrices from vectorized or not connectivities.
```
I cannot understand the end of the sentence: "from vectorized or not connectivities." Do you mean "from connectivities, vectorized or not"?

salma1601 replied:
yes

KamalakerDadi commented on doc/whats_new.rst:
```diff
@@ -1,3 +1,32 @@
+API changes summary
+...................
+
+    - The function `nilearn.connectome.sym_to_vec` is is deprecated and will be removed in future
```
double "is" in this line.

KamalakerDadi replied:
May be we could say just "`sym_to_vec` from connectome module is deprecated ............." ?

Since you are referring a function as `nilearn.connectome.sym_to_vec` which is not included in doc/modules/reference.rst , you will see documentation build error with CircleCI.

GaelVaroquaux commented:
> Ok, I think we should also fix the doc
> https://github.com/nilearn/nilearn/blob/master/nilearn/input_data/
> nifti_maps_masker.py#L87

Yes. Could you please do that. Thanks!

GaelVaroquaux commented:
> May be we could say just "sym_to_vec from connectome module is deprecated

> Since you are referring a function as nilearn.connectome.sym_to_vec which is
> not included in doc/modules/reference.rst , you will see documentation build
> error with CircleCI.

Indeed

KamalakerDadi commented on doc/whats_new.rst:
```diff
@@ -1,3 +1,32 @@
+      connectivity coefficients to connectivity matrices. Also, it allows to
+      recover the covariance matrices for the "tangent" kind.
+
+    - Reworking and renaming of connectivity measures example. 
```
Optional: renaming of connectivity measures example to plot_group_level_connectivity ?

KamalakerDadi commented on examples/03_connectivity/plot_group_level_connectivity.py:
```diff
@@ -0,0 +1,222 @@
+
+plot_matrices(correlation_matrices[:4], 'correlation')
+plotting.plot_connectome(mean_correlation_matrix, msdl_coords,
+                         title='mean correlation over 13 ADHD subjects')
```
For me title looks misleading. It said 13 subjects but we plot only 4. Could we change it to 4 ?

salma1601 replied:
yes buy the average is indeed computed across the 13 subjects

KamalakerDadi replied:
Ok.

KamalakerDadi commented on nilearn/connectome/__init__.py:
```diff
@@ -4,12 +4,13 @@
 
-from .connectivity_matrices import (sym_to_vec, ConnectivityMeasure,
-                                    cov_to_corr, prec_to_partial)
+from .connectivity_matrices import (sym_matrix_to_vec, vec_to_sym_matrix,
```
I think you should put back sym_to_vec to initialize from nilearn.connectome. Since you are deprecating the function and user cannot initialize function directly from nilearn.connectome rather it should be imported from nilearn.connectome.connectivity_matrices. Am I right ?

salma1601 replied:
Ok

GaelVaroquaux commented:
> I think you should put back sym_to_vec to initialize from nilearn.connectome.
> Since you are deprecating the function and user cannot initialize function
> directly from nilearn.connectome rather it should be imported from
> nilearn.connectome.connectivity_matrices. Am I right ?

You are right

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -193,8 +194,40 @@ def _geometric_mean(matrices, init=None, max_iter=10, tol=1e-7):
+        (..., n_features * (n_features + 1) / 2) if discard_diagonal is False and
+        (..., (n_features - 1) * n_features / 2) otherwise.
+    """
+    if discard_diagonal:
```
Can we use function directly here ? Here, we two codes doing the same thing.
```python
return sym_matrix_to_vec(symmetric=symmetric, discard_diagonal=discard_diagonal)
```

salma1601 replied:
what do you mean ?

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -193,8 +194,40 @@ def _geometric_mean(matrices, init=None, max_iter=10, tol=1e-7):
+    If diagonal is kept, diagonal elements are divided by sqrt(2) to conserve
+    the norm.
+    Acts on the last two dimensions of the array if not 2-dimensional.
+    .. versionadded:: 0.2
```
May be we need empty line spaces here

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -193,8 +194,40 @@ def _geometric_mean(matrices, init=None, max_iter=10, tol=1e-7):
     """Return the flattened lower triangular part of an array.
+    If diagonal is kept, diagonal elements are divided by sqrt(2) to conserve
+    the norm.
+    Acts on the last two dimensions of the array if not 2-dimensional.
```
same here emtpy line

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -193,8 +194,40 @@ def _geometric_mean(matrices, init=None, max_iter=10, tol=1e-7):
+        Input array.
+    discard_diagonal : boolean, optional
+        If True, the values of the diagonal are not returned.
+        Default is False.
```
same here empty line

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -193,15 +194,47 @@ def _geometric_mean(matrices, init=None, max_iter=10, tol=1e-7):
 
 
+@deprecated("Function 'sym_to_vec' has been renamed to "
+            "'sym_matrix_to_vec' and will be removed. ")
```
removed + "in future releases".

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -218,19 +220,126 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+
+    Parameters
+    ----------
+    symmetric : numpy.ndarray, shape (..., n_features, n_features)
```
If I understood correctly from tests, this function will accept list of numpy.ndarray ? May be we should update our doc too ?

salma1601 replied:
well it accepts lists but also numpy arrays of shape nD
Do you think I should reformulate the doc? because actually in practice people will use this for a list of 2D matrices (or the equivalent 3D array)

KamalakerDadi commented on nilearn/connectome/connectivity_matrices.py:
```diff
@@ -218,19 +220,126 @@ def sym_to_vec(symmetric, discard_diagonal=False):
+
+    Parameters
+    ----------
+    vec : numpy.ndarray, shape (..., n_columns * (n_columns + 1) /2) or
```
Same here if it accepts list of numpy arrays ?

GaelVaroquaux commented:
> well it accepts lists but also numpy arrays of shape nD
> Do you think I should reformulate the doc? because actually in practice people
> will use this for a list of 2D matrices (or the equivalent 3D array)

Just say that it will also accept a numpy array, or list of numpy arrays.

KamalakerDadi commented:
Sorry for being late.

Can you please update your changes into release section **0.3.1** in ```doc/whats_new.rst``` ?
Thanks

salma1601 commented:
@KamalakerDadi I don't know what to do with the old reference to `sym_to_vec` in whats_new 0.2
It makes circle ci fail. Do I add `sym_to_vec` to the module doc ?

KamalakerDadi commented on doc/modules/reference.rst:
```diff
@@ -40,7 +40,8 @@ uses.
    :toctree: generated/
    :template: function.rst
 
-   sym_to_vec
```
I would suggest to put it back.

The reason being that, this function deprecated but not removed yet. So, if user is still using it (in their pipelines for example) he would understand that this function is renamed and will be removed.

I would stress more on keeping warning message clear and mentioning release name too **0.4**.

KamalakerDadi commented on doc/whats_new.rst:
```diff
@@ -1,13 +1,31 @@
+Changelog
+---------
+
+    - The function sym_to_vec is is deprecated and will be removed in future
```
is is repeated twice. Can you also add release version we want to remove **0.4** ?

KamalakerDadi commented:
@AlexandreAbraham Can you give a call on this ? merge ?

KamalakerDadi commented:
Great work @salma1601 . Merging this one.

GaelVaroquaux commented:
> Great work @salma1601 . Merging this one.

Hurray. Congratulations to everyone involved.

salma1601 commented:
Hurrah !
